### PR TITLE
Fix bug where Sec-WebSocket-Protocol header was required in handshake…

### PR DESCRIPTION
**Pull request by frpermab**
… even if no protocols were given

### CODE REVIEW

File: c/websocket.c [hunk 1]
```diff
@@ -562,12 +562,14 @@ int parse_handshake(ws_ctx_t *ws_ctx, char *handshake) {
         strncpy(headers->connection, start, end-start);
         headers->connection[end-start] = '\0';
    
-        start = strstr(handshake, "\r\nSec-WebSocket-Protocol: ");
-        if (!start) { return 0; }
-        start += 26;
-        end = strstr(start, "\r\n");
-        strncpy(headers->protocols, start, end-start);
-        headers->protocols[end-start] = '\0';
+        if(strlen(headers->protocols)) {
```
Comment (CendioOssman) [inline c/websocket.c:565]: This looks incorrect. This function is supposed to fill out `headers->protocol`, so it doesn't have a sane value before this. So we can't look at it.

File: c/websocket.c [hunk 2]
```diff
@@ -742,7 +743,13 @@ ws_ctx_t *do_handshake(int sock) {
     if (ws_ctx->hybi > 0) {
         handler_msg("using protocol HyBi/IETF 6455 %d\n", ws_ctx->hybi);
         gen_sha1(headers, sha1);
-        snprintf(response, sizeof(response), SERVER_HANDSHAKE_HYBI, sha1, response_protocol);
+        if (!response_protocol) {
```
Comment (CendioOssman) [inline c/websocket.c:746]: You probably need to handle the empty string here as well.

**Review by CendioOssman — CHANGES_REQUESTED** [review]

File: c/websocket.c (hunk 1 above)
Comment (frpermab) [reply c/websocket.c:565]: In fact, headers->protocols is populated at the top of the function, in this line:

headers_t *headers = ws_ctx->headers;

So we *can* look at that field here.

Comment (frpermab) [reply c/websocket.c:565]: Sorry, I should not appear so confident. I *think* this is how it works...

Comment (CendioOssman) [reply c/websocket.c:565]: That only sets `headers`, but not any field within it. If you look at the file you'll see that there is only a single place where `protocols` is assigned any value, and that is the line `strncpy(headers->protocols, start, end-start);`.  So I'm afraid your check cannot look at its value before that.

Comment (frpermab) [reply c/websocket.c:565]: I see. I have made a mistake. I will try to come up with a solution that works properly. Thanks.

**Comment by CendioOssman** [comment]
Closed in favour of #12.

**Pull request closed without merging**